### PR TITLE
EnC: Defer lambda rude edit reporting to runtime

### PR DESCRIPTION
**Pull request by tmat**
Defer the decision of whether or not to allow certain kinds of EnC changes to take effect to the runtime. Apply this approach to lambdas. The IDE currently blocks any changes to lambdas that would alter the shape of the closure (e.g. static lambda changed to a lambda capturing a variable). If, however, the old lambda is not invoked by the application (i.e. the delegate pointing to it is not used after the EnC update) there is no need to block the change. The new lambda will be constructed with the updated closure and can be executed without issues. The IDE does not know for any specific lambda whether or not the application will use the delegate after the edit. However, if we update the old version of the lambda body to throw an exception with rude edit message we can allow the change to go through. If the old lambda is executed it will throw and the debugger will display the rude edit message. As a follow up we will also consider emitting a specific exception type that gets a special treatment from the debugger -- see https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1903644.

Fixes https://github.com/dotnet/roslyn/issues/67841, https://github.com/dotnet/roslyn/issues/68708
Implements https://github.com/dotnet/roslyn/issues/70450, https://github.com/dotnet/roslyn/issues/54672

### CODE REVIEW

**Comment by tmat** [comment]
@davidwengier @cston @jjonescz PTAL

**Review by davidwengier — APPROVED** [review]
EnC specific stuff looks good to me.

File: src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs [hunk 1]
```diff
@@ -20,20 +20,36 @@ namespace Microsoft.CodeAnalysis.Emit
 {
     internal abstract class DefinitionMap
     {
-        protected readonly struct MappedMethod
+        private readonly struct MetadataLambdasAndClosures(
+            ImmutableArray<(DebugId id, IMethodSymbolInternal symbol)> lambdaSymbols,
+            IReadOnlyDictionary<DebugId, (DebugId? parentId, ImmutableArray<string> structCapures)> closureTree)
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs:25]: >Capures [](http://example.com/codeflow?start=90&length=7)

`Captures`?

File: src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs [hunk 2]
```diff
@@ -20,20 +20,36 @@ namespace Microsoft.CodeAnalysis.Emit
...
+            /// <summary>
+            /// Ordered by id.
+            /// </summary>
+            public IReadOnlyDictionary<DebugId, (DebugId? parentId, ImmutableArray<string> structCapures)> ClosureTree { get; } = closureTree;
+
+            public IMethodSymbolInternal? GetLambdaSymbol(DebugId lambdaId)
+                => LambdaSymbols.BinarySearch(lambdaId, static (info, id) => info.id.CompareTo(id)) is int i and >= 0 ? LambdaSymbols[i].symbol : null;
+
+            public (DebugId? parentId, ImmutableArray<string>) TryGetClosureInfo(DebugId closureId)
+                => ClosureTree.TryGetValue(closureId, out var node) ? (node.parentId, node.structCapures) : default;
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs:41]: >(node.parentId, node.structCapures) [](http://example.com/codeflow?start=70&length=35)

Can this be replaced with `node`?

File: src/Compilers/Core/Portable/CodeGen/VariableSlotAllocator.cs [hunk 3]
```diff
@@ -69,17 +70,33 @@ internal abstract class VariableSlotAllocator
         /// <summary>
         /// Finds a closure in the previous generation that corresponds to the specified syntax.
         /// </summary>
+        /// <param name="closureSyntax">Syntax of the closure scope.</param>
+        /// <param name="parentClosureId">Id of the parent closure.</param>
+        /// <param name="structCaptures">Names of variables hoisted into the closure, or empty if the closure is a class.</param>
```
Comment (jjonescz) [inline src/Compilers/Core/Portable/CodeGen/VariableSlotAllocator.cs:75]: Is `structCaptures` sometimes default and sometimes empty? Seems both are mentioned throughout the PR.

File: src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs [hunk 4]
```diff
@@ -522,27 +588,69 @@ IEnumerable<IMethodSymbolInternal> recurse(ImmutableArray<ISymbolInternal> synth
...
+                    id = parsedEntityId;
+                    return true;
+                }
+
+                // invalid metadata
+                id = default;
+                return false;
+            }
+
+            ImmutableArray<string> getHoistedVariableNames(ImmutableArray<ISymbolInternal> members)
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs:642]: >I [](http://example.com/codeflow?start=12&length=1)

Minor: `static`?

File: src/Compilers/Core/Portable/Emit/EditAndContinue/EncMappedMethod.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+using System;
+using Microsoft.CodeAnalysis.Symbols;
+
+namespace Microsoft.CodeAnalysis.Emit;
+
+internal readonly struct EncMappedMethod(IMethodSymbolInternal previousMethod, Func<SyntaxNode, SyntaxNode?>? syntaxMap, Func<SyntaxNode, RuntimeRudeEdit?>? runtimeRudeEdit)
+{
+    public readonly IMethodSymbolInternal PreviousMethod = previousMethod;
+    public readonly Func<SyntaxNode, SyntaxNode?>? SyntaxMap = syntaxMap;
+    public Func<SyntaxNode, RuntimeRudeEdit?>? RuntimeRudeEdit { get; } = runtimeRudeEdit;
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/EncMappedMethod.cs:14]: >{ get; } [](http://example.com/codeflow?start=63&length=8)

Just curious: Why a mix of fields and properties, here and across the new `Enc*Info` and `Enc*Value` types?

**Comment by cston** [comment]
>                         // A new display class and method is generated for lambda that captures x.

Are we capturing `x` in the second generation?

---
Refers to: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs:6602 in 187755b. [](commit_id = 187755b3cf58e42fda4dc0efe220b9424cc112d5, deletion_comment = False)

**Comment by cston** [comment]
>         public void CapureOrdering()

Typo

---
Refers to: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs:8872 in 187755b. [](commit_id = 187755b3cf58e42fda4dc0efe220b9424cc112d5, deletion_comment = False)

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb [hunk 6]
```diff
@@ -912,8 +914,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Private Sub CompileSynthesizedMethods(additionalTypes As ImmutableArray(Of NamedTypeSymbol))
             Debug.Assert(_moduleBeingBuiltOpt IsNot Nothing)
 
-            Dim lambdaDebugInfoBuilder = ArrayBuilder(Of LambdaDebugInfo).GetInstance()
-            Dim closureDebugInfoBuilder = ArrayBuilder(Of ClosureDebugInfo).GetInstance()
+            Dim lambdaDebugInfoBuilder = ArrayBuilder(Of EncLambdaInfo).GetInstance()
+            Dim lambdaRuntimeRudeEditsBuilder = ArrayBuilder(Of LambdaRuntimeRudeEditInfo).GetInstance()
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb:918]: >lambdaRuntimeRudeEditsBuilder [](http://example.com/codeflow?start=16&length=29)

Not calling `Free()`.

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb [hunk 7]
```diff
@@ -1362,8 +1372,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 ' setter needs to rewritten as it may require lambda conversions
                 Dim setterBody = setter.GetBoundMethodBody(compilationState, diagnostics, containingTypeBinder)
 
-                Dim lambdaDebugInfoBuilder = ArrayBuilder(Of LambdaDebugInfo).GetInstance()
-                Dim closureDebugInfoBuilder = ArrayBuilder(Of ClosureDebugInfo).GetInstance()
+                Dim lambdaDebugInfoBuilder = ArrayBuilder(Of EncLambdaInfo).GetInstance()
+                Dim lambdaRuntimeRudeEditsBuilder = ArrayBuilder(Of LambdaRuntimeRudeEditInfo).GetInstance()
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb:1376]: >lambdaRuntimeRudeEditsBuilder [](http://example.com/codeflow?start=20&length=29)

Not calling `Free()`.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 8]
```diff
@@ -1138,8 +1144,9 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                 bool hasBody = flowAnalyzedBody != null;
                 VariableSlotAllocator lazyVariableSlotAllocator = null;
                 StateMachineTypeSymbol stateMachineTypeOpt = null;
-                var lambdaDebugInfoBuilder = ArrayBuilder<LambdaDebugInfo>.GetInstance();
-                var closureDebugInfoBuilder = ArrayBuilder<ClosureDebugInfo>.GetInstance();
+                var lambdaDebugInfoBuilder = ArrayBuilder<EncLambdaInfo>.GetInstance();
+                var lambdaRuntimeRudeEditsBuilder = ArrayBuilder<LambdaRuntimeRudeEditInfo>.GetInstance();
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:1148]: Good catch

File: src/Compilers/Core/Portable/CodeGen/VariableSlotAllocator.cs (hunk 3 above)
Comment (tmat) [reply src/Compilers/Core/Portable/CodeGen/VariableSlotAllocator.cs:75]: Should say "default" here.

File: src/Compilers/Core/Portable/Emit/EditAndContinue/EncMappedMethod.cs (hunk 5 above)
Comment (tmat) [reply src/Compilers/Core/Portable/Emit/EditAndContinue/EncMappedMethod.cs:14]: Fixed. Refactoring added it as a property.

File: src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tmat) [reply src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs:25]: The closure tree elements need to be looked up a couple times during construction, while the lambda symbols are just appended to the array.

**Comment by tmat** [comment]
Had to force push :( Changes are in the last commit.

**Pull request merged**